Initial round of probability distributions and statistical functions (#240)
fortran-lang/stdlib: doc/specs/stdlib_stats_distribution.md

---
title: stats_distribution
---

> # Statistical Distribution

Reviewer: ```suggestion
# Statistical Distributions
```

[TOC]

> ## `random_seed` - set or get a value of seed to the random distribution psudorandom number generator

Reviewer: ```suggestion
## `random_seed` - set or get a value of seed to the random distribution pseudorandom number generator
```

### Status

Experimental

### Description

> Must set the seed value before calling random distribution for variates. 

Reviewer: ```suggestion
Sets the seed value before calling random distribution for variates. 
```

### Syntax

`call random_seed(put [, get])`

### Arguments

> `put`: argument has intent `in` and may be a scalar of type `integer` with kind of int32.

Reviewer: ```suggestion
`put`: argument has intent `in` and may be a scalar of type `integer` with kind of `int32`.
```

> `get`: optional argument has intent `out` and is a scalar of type `integer` with kind of int32.

Reviewer: ```suggestion
`get` (optional): argument has intent `out` and is a scalar of type `integer` with kind of `int32`.
```

> ### Return value

Reviewer: Should be completed to describe `get`

Author: done

### Example

```fortran
program demo_norm_seed
    use stdlib_stats_distribution, only : random_seed
    use iso_fortran_env, only : int32
    implicit none
    integer(int32) :: seed, seed_value

    seed = 1234567
    call random_seed(seed)                   ! reset the value of seed
    call random_seed(seed, seed_value)       ! get current value of seed
end program demo_norm_seed
```

## `uniform_distribution_rvs` - uniform distribution random variates

### Status

Experimental

> ### Description

Reviewer: Would it be possible to return an muti-dimension array (e.g., a ran-3 array), e.g., as in [Matlab](https://www.mathworks.com/help/stats/unifrnd.html)

Author: > Would it be possible to return an muti-dimension array (e.g., a ran-3 array), e.g., as in [Matlab](https://www.mathworks.com/help/stats/unifrnd.html)

Yes. the functions are elemental except the the random number generators, they should work on multi-dimension arrays.

Reviewer: It would be nice that the random number generator functions ("_rvs") also support arrays up to rank 7 or 15 (depending on the compiler). This could be easily generated with Fypp (maybe in a next PR?).

Author: random number generator function itself has no issue with array support, it is elemental. The problem is how to initiate the function call. Normally, one uses one set of parameters to get either one or an array of random numbers. This is how current implementation worked. If we want array support up to rank 15, then we need a mask array as an optional argument to the function call. If peoples like the idea, we can certainly implement it next round.

Reviewer: I was thinking to extend the rank-1 function to multiple ranks, e.g. with multiple loops. Therefore, there is no need for optional arguments (there would rank+1 different procedures). But, as you mentioned, we can think about that in a next round if desired.

Using the parameters `loc` and `scale`, one obtains the uniformly distributed random variates on [loc, loc + scale]. Return a scalar of type `real` pseudorandom number or a rank one array of such numbers with a size of `array_size`.

### Syntax

`result = [[stdlib_stats_distribution(module):uniform_distribution_rvs(interface)]](loc, scale, [array_size])`

### Arguments

`array_size`: optional argument has intent `in` and is a scalar of type `integer`.

`loc`: has intent `in` and is a scalar of type `real`.

`scale`: has intent `in` and is a scalar of type `real`.

### Return value

The result is a scalar of type `real`, or is a rank one array of type `real` with size of `array_size`.

### Example

```fortran
program demo_uniform_rvs
    use stdlib_stats_distribution, only: uniform => uniform_distribution_rvs
    implicit none
    real :: loc, scale, rv(10)

    loc = 0.0
    scale = 1.0

>     print *, uniform(loc, scale)            ! single standard uniform random variate in [0, 1)

Reviewer: ```suggestion
    print *, uniform(loc, scale)            ! single standard uniform random variate in [0, 1]
```

>     rv = uniform(loc, scale, 10)            ! an array of 10 uniform random variates in [0, 1)

Reviewer: ```suggestion
    rv = uniform(loc, scale, 10)            ! an array of 10 uniform random variates in [0, 1]
```

    loc = -1.0
    scale = 2.0

>     print *, uniform(loc, scale)            ! single uniform random variate in [-1,1)

Reviewer: ```suggestion
    print *, uniform(loc, scale)            ! single uniform random variate in [-1,1]
```

    loc = -0.5
    scale = 1.0

>     rv = uniform(loc, scale, 10)            ! an array of 10 uniform variates in [-0.5, 0.5)

Reviewer: ```suggestion
    rv = uniform(loc, scale, 10)            ! an array of 10 uniform variates in [-0.5, 0.5]
```

end program demo_uniform_rvs
```

## `uniform_distribution_pdf` - uniform probability density function

### Status

Experimental

### Description

The probability density function of the continuous uniform distribution.

> ![equation](https://latex.codecogs.com/gif.latex?f(x)=\begin{cases}\frac{1}{scale}&loc\leqslant&space;x<loc&plus;scale\\\\0&x<loc,or,x>loc&plus;scale\end{cases})

Reviewer: FORD supports Latex: see [here](https://github.com/Fortran-FOSS-Programmers/ford/wiki/Writing-Documentation#latex-support) for more details.
Something like (I didn't check the formula):
```suggestion
\( \frac{1}{scale}&loc\leqslant&space;x<loc&plus;scale\\\\0&x<loc,or,x>loc&plus;scale )\
```

Author: > FORD supports Latex: see [here](https://github.com/Fortran-FOSS-Programmers/ford/wiki/Writing-Documentation#latex-support) for more details.
> Something like (I didn't check the formula):

I tried FORD LaTex support, It seems not working for my formulae.

Reviewer: Ok. Thanks for trying. I will try it later too. IMO it is preferable to not rely on a third party app.

### Syntax

`result = [[stdlib_stats_distribution(module):uniform_distribution_pdf(interface)]](x, loc, scale)`

### Arguments

`x`: has intent `in` and is a scalar or an array of type `real`.

`loc`: has intent `in` and is a scalar of type `real`.

`scale`: has intent `in` and is a scalar of type `real`.

### Return value

The result is a scalar of type `real` or an array of type `real` with a shape conformable to input `x`.

### Example

```fortran
program demo_uniform_pdf
    use stdlib_stats_distribution,  uniform_pdf => uniform_distribution_pdf, &
                                    uniform => uniform_distribution_rvs
    implicit none
    real :: loc, scale, x(3,4)

    loc = 0.0
    scale = 1.0

>     print *, uniform_pdf(0.5, loc, scale)             ! a probability density at 0.5 in [0,1)
>     print *, uniform_pdf(0.7, -1.0, 2.0)              ! a probability density at 0.7 in [-1,1)
>     x = reshape(uniform(loc, scale, 12),[3,4])        ! uniform random variates array in [0,1)
>     print *, uniform_pdf(x, loc, scale)               ! probability density array in [0,1)

Reviewer: ```suggestion
    print *, uniform_pdf(0.5, loc, scale)             ! a probability density at 0.5 in [0,1]
    print *, uniform_pdf(0.7, -1.0, 2.0)              ! a probability density at 0.7 in [-1,1]
    x = reshape(uniform(loc, scale, 12),[3,4])        ! uniform random variates array in [0,1]
    print *, uniform_pdf(x, loc, scale)               ! probability density array in [0,1]
```

end program demo_uniform_pdf

```

## `uniform_distribution_cdf` - uniform cumulative distribution function

### Status

Experimental

### Description

Cumulative distribution function of the uniform continuous distribution

![equation](https://latex.codecogs.com/gif.latex?F(x)=\begin{cases}\0&x<loc\\\\\frac{x-loc}{scale}&loc\leq&space;x<loc&plus;scale\\\\1&x\geq&space;loc&plus;scale\end{cases})

### Syntax

`result = [[stdlib_stats_distribution(module):uniform_distribution_cdf(interface)]](x, loc, scale)`

### Arguments

`x`: has intent `in` and is a scalar or an array of type `real`.

`loc`: has intent `in` and is a scalar of type `real`.

`scale`: has intent `in` and is a scalar of type `real`.

### Return value

The result is a scalar of type `real` or an array of type `real` with a shape conformable to input `x`.

### Example

```fortran
program demo_uniform_cdf
    use stdlib_stats_distribution, uniform_cdf => uniform_distribution_cdf, &
                                   uniform => uniform_distribution_rvs
    implicit none
    real :: loc, scale, x(3,4)

loc = 0.0
    scale = 1.0

>     print *, uniform_cdf(0.5, loc, scale)             ! a cumulative at 0.5 in [0,1)
>     print *, uniform_cdf(0.7, -1.0, 2.0)              ! a cumulative at 0.7 in [-1,1)
>     x = reshape(uniform(loc, scale, 12),[3,4])        ! uniform random variates array in [0,1)
>     print *, uniform_cdf(x, loc, scale)               ! cumulative array in [0,1)

Reviewer: ```suggestion
    print *, uniform_cdf(0.5, loc, scale)             ! a cumulative at 0.5 in [0,1]
    print *, uniform_cdf(0.7, -1.0, 2.0)              ! a cumulative at 0.7 in [-1,1]
    x = reshape(uniform(loc, scale, 12),[3,4])        ! uniform random variates array in [0,1]
    print *, uniform_cdf(x, loc, scale)               ! cumulative array in [0,1]
```
